zebra-rpc/src/queue.rs

```diff
@@ -0,0 +1,323 @@
+//! Transaction Queue.
+//!
+//! All transactions that are sent from RPC methods should be added to this queue for retries.
+//! Transactions can fail to be inserted to the mempool inmediatly by different reasons,
+//! like having not mined utxos.
+//!
+//! The [`Queue`] is just an `IndexMap` of transactions with insertion date.
+//! We use this data type because we want the transactions in the queue to be in order.
+//! The [`Runner`] component will do the processing in it's [`Runner::run()`] method.
+
+use std::{collections::HashSet, sync::Arc};
+
+use chrono::Duration;
+use indexmap::IndexMap;
+use tokio::{
+    sync::broadcast::{channel, Receiver, Sender},
+    time::Instant,
+};
+
+use tower::{Service, ServiceExt};
+
+use zebra_chain::{
+    block::Height,
+    chain_tip::ChainTip,
+    parameters::{Network, NetworkUpgrade},
+    transaction::{Transaction, UnminedTx, UnminedTxId},
+};
+use zebra_node_services::{
+    mempool::{Gossip, Request, Response},
+    BoxError,
+};
+
+use zebra_state::{ReadRequest, ReadResponse};
+
+#[cfg(test)]
+mod tests;
+
+/// The number of blocks a transaction can be in the queue.
+const NUMBER_OF_BLOCKS_TO_EXPIRE: i64 = 3;
+
+/// Size of the queue and channel. Suggested value is equal to
+/// `mempool::downloads::MAX_INBOUND_CONCURRENCY`
+const CHANNEL_AND_QUEUE_CAPACITY: usize = 10;
```

Comment:
We want to allow room for new transactions or existing mempool transactions in the queue.
(We add transactions from the channel, then check if they need to be removed.)

```suggestion
const CHANNEL_AND_QUEUE_CAPACITY: usize = 20;
```

Reply:
https://github.com/ZcashFoundation/zebra/pull/4015/commits/4afcfba4c28cfb83407ede3c19779fe52a49700a